### DISCUSSION TITLE
how to use pymudpdf to convert html (with page break div and customized font ttf font fact) or epub to pdf?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#2114] [Q&A] [answered]
how to use pymudpdf to convert html (with page break div and customized font ttf font fact) or epub to pdf?

**Comment by JorjMcKie** [comment]
This depends on some details of your situation. You may want to have look at the features around the new `Story` class.
This feature is introduced in the [tutorial](https://pymupdf.readthedocs.io/en/latest/tutorial.html#workingwithstories) and further explained in its own chapter within the [recipes](https://pymupdf.readthedocs.io/en/latest/recipes-stories.html) chapter.
The basic approach is:
* Given HTML and optional CSS sources, invoke `Story` methods to produce pages of a new PDF, filled with the interpreted HTML text - styled according to embedded styling instructions or via a provided CSS source.
* This PDF page production can be programmatically influenced in various ways, e.g. by providing desired page size(s), fillable rectangle(s) on each page, and more
* Before actually starting the PDF production, the HTML may also be manipulated in pretty arbitrary ways from within your script: deleting, adding, changing content or styling, automatic Table of Contents creation, etc.

Because HTML / CSS are being used as specification language, you can do pretty much most of what is possible with HTML, e.g. also using your own fonts via `@font-face`.

**Comment by nissansz** [comment]
Thank you. Any direct method like calibre? Input an html/epub address, output an pdf.

**Comment by nissansz** [comment]
Below code works. 
But prints one error: mupdf: cannot read archive entry.
Also the exported file is very large, html is 100 kb. pdf is 3MB. And don't know where to add ", garbage=3,deflate=True)"


import pathlib
import fitz

htmlpath = pathlib.Path("myhtml.html")
# csspath = pathlib.Path("mycss.css")

HTML = htmlpath.read_bytes().decode()
# CSS = csspath.read_bytes().decode()

# story = fitz.Story(html=HTML, user_css=CSS)
story = fitz.Story(html=HTML)

writer = fitz.DocumentWriter("output.pdf")  # create the writer
# MEDIABOX = fitz.paper_rect("letter")  # output page format: Letter
MEDIABOX = fitz.paper_rect("A3")  # output page format: Letter
WHERE = MEDIABOX + (36, 36, -36, -36)  # leave borders of 0.5 inches
more = 1  # will indicate end of input once it is set to 0

while more:  # loop outputting the story
    device = writer.begin_page(MEDIABOX)  # make new page
    more, _ = story.place(WHERE)  # layout into allowed rectangle
    story.draw(device)  # write on page
    writer.end_page()  # finish page

writer.close()  # close output file

**Comment by nissansz** [comment]
base64 image in html code is missing in pdf. Local font did not effect.
<img width="65" height="42" src="data:image/jpeg;base64,

	        @font-face {
            font-family: swf-f0f95;
            src: url(f0f95.ttf);
        }

**Comment by JorjMcKie** [comment, marked as answer]
1. Images embedded in the HTML (as base64) are **_not supported_**.
2. If referring to font or image resources you must use the `Archive` class - a generalized container concept: you can mix file folders and ZIP or TAR archives in one `fitz.Archive`.
3. When defining the `fitz.DocumentWriter` you can use "compress" to reduce the resulting PDF file size. In addition, instead of a file, you can let DpcumentWriter write to a Python file pointer and reopen the resulting memory PDF for your own post processing. This is an example which builds font subsets and other methods for size reduction:

```python
import io
import fitz

fileptr = io.BytesIO()
writer = fitz.DocumentWriter(fileptr)
# story processing here ...
...
writer.close()
doc = fitz.open("pdf", fileptr) # reopen memory PDF
doc.subset_fonts() # build subsets for eligible fonts
doc.ez_save("myfile.pdf") # save using compression and duplicate object elimination
```

**Comment by nissansz** [comment]
It works to compress file size, now pdf size is 153kb.
Any solution for images? Convert base64 to local image or memory, then insert as html code?
Any sample code about Archive font for pdf?

**Reply by JorjMcKie** [reply]
The HTML tag for images `<img>` is supported. You must write your own code to extract images embedded in your HTML and store them outside as regular image files.

> Any sample code about Archive font for pdf?

You need to write CSS code for pointing to own fonts - either in a separate CSS source or embedded inside your HTML.
If you need an example how this may look like I suggest to read CSS / HTML documentation **_or_** use the output of a new method as a guideline.
The following code generates font support for the "ubuntu" fonts inside package pymupdf-fonts:
```python
arch=fitz.Archive()
css = fitz.css_for_pymupdf_font("ubuntu", archive=arch)
print(css)

@font-face {font-family: ubuntu; src: url(ubuntu);}

@font-face {font-family: ubuntu; src: url(ubuntubo);font-weight: bold;}

@font-face {font-family: ubuntu; src: url(ubuntubi);font-weight: bold;font-style: italic;}

@font-face {font-family: ubuntu; src: url(ubuntuit);font-style: italic;}

# define the story like this:
story = fitz.Story(html, user_css=css, archive=arch)
```
Inside the HTML, the font-family "ubuntu" can be used and will take the appropriate regular, italic, etc. variant.
If you use `fitz.css_for_pymupdf_font(..., name="sans-serif")` then all references to the HTML standard font family "sans-serif" will automatically use ubuntu instead. Similar for "serif" or "monospace".

If you have your own font file, create similar `@font-face` syntax and refer to the fontfile via the url(). Your archive must then also point to the local folder where the fontfile lives: `arch.add("foldername")`.

**Comment by nissansz** [comment]
Thank you. I tried chrome headless, it seems fine to work.
